x/stake/simulation/invariants.go

```diff
@@ -131,6 +141,53 @@ func PositivePowerInvariant(k stake.Keeper) simulation.Invariant {
 	}
 }
 
+// PositiveDelegationInvariant checks that all stored delegations have > 0 shares.
+func PositiveDelegationInvariant(k stake.Keeper) simulation.Invariant {
+	return func(app *baseapp.BaseApp) error {
+		ctx := app.NewContext(false, abci.Header{})
+
+		delegations := k.GetAllDelegations(ctx)
+		for _, delegation := range delegations {
+			if delegation.Shares.IsNegative() {
+				return fmt.Errorf("delegation with negative shares: %+v", delegation)
+			}
+			if delegation.Shares.IsZero() {
+				return fmt.Errorf("delegation with zero shares: %+v", delegation)
+			}
+		}
+
+		return nil
+	}
+}
+
+// DelegatorSharesInvariant checks whether all the delegator shares which persist
+// in the delegator object add up to the correct total delegator shares
+// amount stored in each validator
+func DelegatorSharesInvariant(k stake.Keeper) simulation.Invariant {
```

Comment:
Also a computationally expensive invariant, maybe it should be periodic by default